src/extensions/yfm/GPT/MarkupGpt/plugin.ts

```diff
@@ -0,0 +1,165 @@
+import {EditorState} from '@codemirror/state';
+import {WidgetType} from '@codemirror/view';
+
+import {GptWidgetOptions} from '../../..';
+import {
+    Decoration,
+    type DecorationSet,
+    type EditorView,
+    type PluginValue,
+    ViewPlugin,
+    type ViewUpdate,
+} from '../../../../cm/view';
+import {ReactRendererFacet} from '../../../../markup';
+
+import {hideMarkupGptExample} from './commands';
+import {HideMarkupGptExampleEffect, ShowMarkupGptExampleEffect} from './effects';
+import {renderPopup} from './popup';
+
+const DECO_CLASS_NAME = 'react-popup-example-deco';
+
+class SpanWidget extends WidgetType {
+    private className = '';
+    private textContent = '';
+
+    constructor(className: string, textContent: string) {
+        super();
+        this.className = className;
+        this.textContent = textContent;
+    }
+
+    toDOM() {
+        const spanElem = document.createElement('span');
+        spanElem.className = this.className;
+        spanElem.textContent = this.textContent;
+        return spanElem;
+    }
+}
+
+export function mGptExamplePlugin(gptProps: GptWidgetOptions) {
+    return ViewPlugin.fromClass(
+        class implements PluginValue {
+            readonly _view: EditorView;
+            readonly _renderItem;
+
+            decos: DecorationSet = Decoration.none;
+            _anchor: Element | null = null;
+            disablePromptPresets = true;
+            markup: string | null = null;
+            state: null | EditorState = null;
+
+            selectedPosition = {
+                from: 0,
+                to: 0,
+            };
+
+            constructor(view: EditorView) {
+                this._view = view;
+                this._renderItem = view.state
+                    .facet(ReactRendererFacet)
+                    .createItem('react-popup-example-in-markup-mode', () => this.renderPopup());
+            }
+
+            update(update: ViewUpdate) {
+                if (update.docChanged || update.selectionSet) {
+                    this.decos = Decoration.none;
+                    return;
+                }
+
+                this.decos = this.decos.map(update.changes);
+                this.state = update.state;
```

Comment:
It is unnecessary to save the entire state, it is enough to save only `from` and `to` from the selection